cloudscape-design/components
# fix: Keep chart popover position on resize

### Description

Splitting #1644 in smaller parts.

The popover container recalculates their position when any of this happens:
- on scroll
- on page resize
- on popover resize
- on click (anywhere on the page)

However, when a click comes from inside the popover itself, this can lead to situations where the popover moves away from the cursor. This can be especially critical with chart popovers, because they are shown on hover, so that clicking on them can lead to the actual popover disappearing because the hover was lost:

https://github.com/cloudscape-design/components/assets/1257272/7175af27-30fd-47c3-9f9d-e0480c42a0e4

After these changes, the popover position is kept:

https://github.com/cloudscape-design/components/assets/1257272/b43b86ea-8602-49a3-b21b-c853ffb7daf8

In order to reduce any possible unexpected impact on other popovers, these changes are applied here only on chart popovers, via a new prop `keepPosition` in the internal popover container component.

### How has this been tested?

- Added integration tests
- Added unit tests
- Manually tested on the new dev page
- Manually tested on a website dry-run playground to make sure it also works on React 18

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
